src/libcore/str/mod.rs

```diff
@@ -1757,9 +1757,9 @@ mod traits {
         }
         #[inline]
         unsafe fn get_unchecked_mut(self, slice: &mut str) -> &mut Self::Output {
-            let ptr = slice.as_ptr().add(self.start);
+            let ptr = slice.as_mut_ptr().add(self.start);
             let len = self.end - self.start;
-            super::from_utf8_unchecked_mut(slice::from_raw_parts_mut(ptr as *mut u8, len))
+            super::from_utf8_unchecked_mut(slice::from_raw_parts_mut(ptr, len))
```

Comment:
Notice that this raw ptr cast here is the "canary" that gave away that the old code was wrong -- it was actually casting `*const u8` to `*mut u8`, which should not have been necessary.